File: shared/types/screenshot.ts
/**
 * Screenshot AI contract — Phase 19 Step 2.
 *
 * ⚠️ AI OBSERVES. CORE DECIDES.
 *
 * Everything in this file is shaped by one rule: vision output is EVIDENCE, and
 * evidence is not a decision. There is no field here for priority, severity,
 * routing, assignment, status, SLA, resolution or a customer reply — not
 * "optional" ones, not nullable ones, none. A model that wanted to set a
 * ticket's severity has nowhere to write it, so the boundary is enforced by the
 * shape of the type rather than by a reviewer noticing.
 *
 * That is the same technique Phase 12 and 13 used to stop a model naming a
 * document Core did not supply: make the wrong thing unrepresentable instead of
 * merely invalid.
 *
 * ⚠️ NO NODE IMPORTS. Imported by the admin panel through the
 * `@iris/shared/screenshot` subpath, exactly as the KB lifecycle is, so that
 * `node:crypto` (via ids.ts in the types barrel) never reaches the browser
 * bundle.
 */

// ─────────────────────────────────────────────────────────────────────────
// What may be analysed
// ─────────────────────────────────────────────────────────────────────────

/**
 * ⚠️ NARROWER THAN THE UPLOAD ALLOWLIST, DELIBERATELY.
 *
 * `image/gif` is a valid attachment and is content-verified at upload, but it
 * is NOT analysed. An animation raises a question this phase does not answer —
 * which frame is the evidence? — and answering it silently by taking the first
 * frame would make the interpretation depend on something nobody chose. PDF is
 * absent for the same class of reason: it is a document that would need
 * rendering IRIS does not do.
 *
 * Anything not on this list is skipped as a PERMANENT, expected outcome, not a
 * failure: attaching a CSV to a ticket is normal, and it must not produce an
 * error anyone has to triage.
 */
export const SCREENSHOT_MIME_TYPES = ['image/png', 'image/jpeg', 'image/webp'] as const;
export type ScreenshotMimeType = (typeof SCREENSHOT_MIME_TYPES)[number];

export function isScreenshotMimeType(value: string): value is ScreenshotMimeType {
  return (SCREENSHOT_MIME_TYPES as readonly string[]).includes(value);
}

/**
 * ⚠️ THE VISION DISPATCH BOUND, WHICH IS NOT THE STORAGE BOUND.
 *
 * Phase 19 Step 1 set MAX_IMAGE_WIDTH/HEIGHT 20000 and 50 MP. Those are
 * SAFETY limits chosen against decompression bombs, and they run on every image
 * upload including ones no model will ever see, so they are deliberately loose
 * enough not to reject a real full-page capture.
 *
 * These are different limits for a different reason: cost and usefulness. A
 * vision model tiles an image and stops gaining signal past roughly 2000px on
 * the short side, so a 40 MP screenshot is billed as a large image and read as
 * a small one. Bytes are the operative bound because bytes are what crosses the
 * wire, inflate by 4/3 as base64, and land in the provider's request body.
 *
 *   4 MiB   comfortably holds a lossless 4K PNG screenshot (typically 1-3 MiB)
 *           and becomes ~5.3 MiB of base64 in the JSON body.
 *   12000   per axis, and 40 MP, are the secondary guard for the pathological
 *           case of a huge canvas that still compresses small.
 *
 * ⚠️ APPLIED IN CORE, BEFORE THE WORKER IS GIVEN THE BYTES. Rejecting at
 * dispatch rather than at the provider is what makes an oversized screenshot
 * cost nothing: no HTTP call, no tokens, no bill. It is also why the check
 * cannot live in Python — by then the payload has already been sent.
 */
export const SCREENSHOT_MAX_BYTES = 4 * 1024 * 1024;
export const SCREENSHOT_MAX_WIDTH = 12_000;
export const SCREENSHOT_MAX_HEIGHT = 12_000;
export const SCREENSHOT_MAX_PIXELS = 40_000_000;

// ─────────────────────────────────────────────────────────────────────────
// The structured output
// ─────────────────────────────────────────────────────────────────────────

/**
 * What kind of thing was seen.
 *
 * A CLOSED SET, and small on purpose. An open `type` string would let the model
 * invent a taxonomy, and a second taxonomy alongside the product's configured
 * categories is exactly what the Phase 19 audit ruled out: Screenshot AI
 * supplies evidence that the EXISTING classification can consume, and never a
 * competing category system.
 */
export const SCREENSHOT_OBSERVATION_TYPES = [
  'error_code',
  'error_message',
  'application',
  'ui_element',
  'other',
] as const;
export type ScreenshotObservationType = (typeof SCREENSHOT_OBSERVATION_TYPES)[number];

/**
 * Bounds. Every one of these is enforced by Core, not merely requested in a
 * prompt — the model is untrusted output, so "the schema said 10" is a hope and
 * a validator is a guarantee.
 *
 * They are small because the consumer is an agent glancing at a ticket, and
 * because an unbounded array is an unbounded row in `ai_execution.result`.
 */
export const SCREENSHOT_MAX_OBSERVATIONS = 10;
export const SCREENSHOT_MAX_OBSERVATION_CHARS = 200;
export const SCREENSHOT_MAX_HINT_CHARS = 60;
export const SCREENSHOT_MAX_CAUSES = 5;
export const SCREENSHOT_MAX_STEPS = 5;
export const SCREENSHOT_MAX_TEXT_CHARS = 300;

/**
 * A hard ceiling on the whole validated payload, checked after field
 * validation. The per-field bounds already imply a maximum, so this exists to
 * catch the case they do not: a future field added without a bound. Cheap, and
 * it fails loudly rather than growing `ai_execution.result` quietly.
 */
export const SCREENSHOT_MAX_RESULT_BYTES = 8 * 1024;

export interface ScreenshotObservation {
  type: ScreenshotObservationType;
  /** What was literally visible. Never an inference. */
  value: string;
}

/**
 * A HINT, and the name is load-bearing.
 *
 * These are free strings rather than the product's category enum, and that is
 * deliberate in both directions: the model must not be handed the taxonomy to
 * choose from (that is classification's job, and it re-validates against Core's
 * own vocabulary), and it must not be able to assert a category that looks
 * authoritative. A hint is something a human or a later classification pass can
 * weigh; it is not a label.
 */
export interface ScreenshotProblemHint {
  domain: string | null;
  category: string | null;
}

/**
 * ⚠️ NOTE WHAT IS ABSENT AND MUST STAY ABSENT: priority, severity, category
 * (as an authoritative field), assignee, team, routing, status, sla, resolution,
 * customer_message, action, auto_*. None of these has a field here, so vision
 * cannot express a decision even if a future prompt asked it to.
 */
export interface ScreenshotInterpretation {
  observations: ScreenshotObservation[];
  problem_hint: ScreenshotProblemHint;
  possible_causes: string[];
  suggested_next_steps: string[];
  /**
   * ⚠️ THE MODEL'S OWN REPORTED SIGNAL. NOT A CALIBRATED PROBABILITY.
   *
   * Phase 4 measured exactly this on classification: the model returned >=0.95
   * on almost every ticket, which is high self-reported confidence and says
   * nothing about whether the signal discriminates correct from incorrect.
   * Store it, show it labelled as what it is, and never render it as an
   * accuracy figure. See the UI rules in the admin page.
   */
  confidence: number;
}

/**
 * The keys Core will accept. Anything else is a rejection, not a silent drop.
 *
 * Exported so the validator and its tests share one definition: a field added
 * to the interface but not here would be stripped, and a field added here but
 * not validated would be persisted unchecked.
 */
export const SCREENSHOT_RESULT_KEYS = [
  'observations',
  'problem_hint',
  'possible_causes',
  'suggested_next_steps',
  'confidence',
] as const;

/**
 * Field names that would represent a DECISION rather than evidence.
 *
 * ⚠️ THIS IS A TRIPWIRE, NOT THE ENFORCEMENT. The enforcement is that the
 * accepted key set above is closed, so every one of these is already rejected
 * as an unexpected field. This list exists so that a rejection carrying one of
 * these names is reported distinctly: a model returning `severity` is not a
 * schema typo, it is the model trying to make a business decision, and that is
 * worth being able to see in the logs.
 */
export const SCREENSHOT_FORBIDDEN_KEYS = [
  'priority',
  'severity',
  'assignee',
  'assignee_id',
  'team',
  'routing',
  'route_to',
  'status',
  'sla',
  'resolution',
  'resolved',
  'customer_message',
  'reply',
  'comment',
  'action',
  'auto_resolve',
] as const;

// ─────────────────────────────────────────────────────────────────────────
// What Core stores and the UI reads
// ─────────────────────────────────────────────────────────────────────────

/**
 * One screenshot interpretation as the admin panel receives it.
 *
 * `attachment_id` is carried in the VALIDATED RESULT rather than in a new
 * `ai_execution` column. The Phase 19 audit called this out: the linkage
 * already exists through the event, a column would be a migration, and the id
 * is an opaque ULID that identifies no tenant. Core writes it from the event
 * payload it resolved itself; the model has no field in which to influence it.
 */
export interface ScreenshotResultDTO {
  execution_id: string;
  attachment_id: string | null;
  status: 'running' | 'succeeded' | 'failed';
  interpretation: ScreenshotInterpretation | null;
  error_code: string | null;
  model: string | null;
  provider: string | null;
  latency_ms: number | null;
  created_at: string;
  completed_at: string | null;
}

File: ai-service/src/api/screenshot.py
"""Screenshot interpretation — Phase 19.

The first multimodal feature. One already-authorized image plus the ticket text
in, structured visual evidence out.

⚠️ AI OBSERVES. CORE DECIDES.

`ScreenshotOutput` has no field for a priority, severity, urgency, impact,
assignee, team, queue, routing, status, SLA, resolution or customer message.
That is not politeness enforced by prompt wording — there is nowhere to put one,
so a model that tries fails this schema, and Core rejects the same payload again
independently. The prompt below exists so the model does not waste a paid call
producing a response that could only be thrown away.

⚠️ THIS SERVICE NEVER PERSISTS OR LOGS THE IMAGE. It holds no database handle
and writes nothing to disk; the bytes exist for one provider call. Nothing in
this module logs `req.input.image`, and the test suite asserts no base64 reaches
a log line.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .prompts import TICKET_DELIMITER

#: Bumped when the wording changes in a way that could move outputs. Persisted
#: on every execution, so a shift in behaviour is attributable to a prompt
#: change rather than guessed at.
SCREENSHOT_PROMPT_VERSION = "screenshot-v1"


class ScreenshotObservation(BaseModel):
    """One thing literally visible in the image.

    `type` is a closed set. An open string would let the model invent a
    taxonomy, and a second taxonomy alongside the product's configured
    categories is exactly what the Phase 19 audit ruled out: Screenshot AI
    supplies evidence the EXISTING classification can consume, never a competing
    category system.
    """

    model_config = ConfigDict(extra="forbid")

    type: Literal["error_code", "error_message", "application", "ui_element", "other"]
    value: str = Field(
        min_length=1,
        max_length=200,
        description="The text or element exactly as it appears in the image.",
    )


class ScreenshotProblemHint(BaseModel):
    """A hint for a human, and the name is load-bearing.

    Free strings rather than the product's category enum, deliberately in both
    directions: the model is not handed a vocabulary to choose from (that is
    classification's job, and Core re-validates it against its own taxonomy),
    and it cannot assert a category that would read as authoritative.

    Both fields nullable, and BOTH NULL IS A GOOD ANSWER. A screenshot showing
    nothing recognisable should say so; forcing a guess is how a hint becomes
    noise an agent learns to skip.
    """

    model_config = ConfigDict(extra="forbid")

    domain: str | None = Field(default=None, max_length=60)
    category: str | None = Field(default=None, max_length=60)


class ScreenshotOutput(BaseModel):
    """The entire contract.

    Bounds mirror shared/types/screenshot.ts. Core re-validates every one of
    them: this service accepting a payload has never been a security property.
    """

    model_config = ConfigDict(extra="forbid")

    observations: list[ScreenshotObservation] = Field(default_factory=list, max_length=10)
    problem_hint: ScreenshotProblemHint
    possible_causes: list[str] = Field(default_factory=list, max_length=5)
    suggested_next_steps: list[str] = Field(default_factory=list, max_length=5)
    confidence: float = Field(
        ge=0,
        le=1,
        description=(
            "How much the IMAGE supports this interpretation. Self-reported, not "
            "calibrated; a blurry or cropped screenshot must score low."
        ),
    )


#: ⚠️ THE PROMPT'S JOB IS TO STOP THE MODEL DECIDING THINGS.
#:
#: A vision model shown a support screenshot will volunteer a severity, a team
#: to route to and a reply to send, because the surrounding context implies that
#: is what a support system wants. None of those has a field in the schema, so
#: such a response is rejected in full — the prompt exists so the model does not
#: spend a call producing one.
#:
#: The separation between OBSERVED and INFERRED is the substance of the feature.
#: An agent can act on "the banner reads ERR_QUOTA_EXCEEDED"; they cannot act on
#: a fluent guess that is phrased identically. So observations are literal, and
#: everything speculative is confined to two clearly-named lists.
SCREENSHOT_SYSTEM_PROMPT = """You are a support engineer examining ONE screenshot attached to a support ticket.

Your only job is to report what is VISIBLE and to offer cautious, clearly-labelled interpretation. You are gathering evidence for a human agent. You are not diagnosing, deciding or replying.

WHAT TO RECORD AS AN OBSERVATION
- Only things literally visible in the image. Transcribe error codes and error messages exactly as shown, character for character.
- If text is cut off or unreadable, do not complete it. Omit it.
- Do not describe an element that is not there. An empty observations list is a valid and useful answer.
- Do not name the product or vendor from appearance alone. Report an application name only if it is written in the image.

INTERPRETATION
- possible_causes and suggested_next_steps are explicitly speculative. Keep them short, concrete and few.
- If the screenshot does not support a conclusion, return fewer items, or none. Do not fill the lists to look thorough.
- problem_hint is a hint for a human, not a classification. Use null for domain or category when the image does not clearly indicate one.

CONFIDENCE
- Report your own confidence between 0 and 1, reflecting how much the IMAGE supports your interpretation.
- A blurry, cropped or ambiguous screenshot must score low. Do not report high confidence out of politeness.

SENSITIVE CONTENT
- Screenshots often contain personal or secret data. Never transcribe a password, API key, token, session id, bearer value, card number or anything that looks like a credential, even when clearly visible.
- Do not transcribe email addresses, usernames, account numbers or internal URLs unless they form part of an error message and are necessary to understand it.
- Reproduce the minimum text needed to make the problem identifiable.

WHAT YOU MUST NOT DO
- Do not assign or suggest a priority, severity, urgency, impact rating, team, queue, assignee, status or resolution.
- Do not write a message to the customer.
- Do not state that the ticket should be closed, escalated or resolved.
- Those decisions belong to the support platform and are made elsewhere. There is no field for them, and a response containing one is rejected in full.

Reply with ONLY the JSON object required by the schema."""


def build_screenshot_user_prompt(*, subject: str | None, description: str) -> str:
    """The ticket text that accompanies the image.

    The customer's own words are included because a screenshot read without them
    is a caption exercise: "a red banner" means something different on a ticket
    about failed exports than on one about login.

    Same delimiters and the same absence of identifiers as every other prompt
    here — no tenant, no ticket id, no product name, none of which this service
    is given (see ExecuteRequest in schemas.py).
    """
    subject_line = (subject or "").strip() or "(no subject)"
    return (
        "The customer wrote the following, and attached the image below.\n\n"
        f"{TICKET_DELIMITER}\n"
        f"Subject: {subject_line}\n"
        f"Description: {description.strip()}\n"
        f"{TICKET_DELIMITER}\n\n"
        "Examine the attached screenshot and report what is visible."
    )
